src/cdomains/apron/sharedFunctions.apron.ml

```diff
@@ -75,81 +80,123 @@ struct
-            | BinOp (Mult, e1, e2, _) ->
-              Binop (Mul, texpr1_expr_of_cil_exp e1, texpr1_expr_of_cil_exp e2, Int, Near)
-            | BinOp (Div, e1, e2, _) ->
-              Binop (Div, texpr1_expr_of_cil_exp e1, texpr1_expr_of_cil_exp e2, Int, Zero)
-            | BinOp (Mod, e1, e2, _) ->
-              Binop (Mod, texpr1_expr_of_cil_exp e1, texpr1_expr_of_cil_exp e2, Int, Near)
-            | CastE (TInt (t_ik, _) as t, e) ->
-              begin match IntDomain.should_ignore_overflow t_ik || IntDomain.Size.is_cast_injective ~from_type:(Cilfacade.typeOf e) ~to_type:t with (* TODO: unnecessary cast check due to overflow check below? or maybe useful in general to also assume type bounds based on argument types? *)
-                | true ->
-                  Unop (Cast, texpr1_expr_of_cil_exp e, Int, Zero) (* TODO: what does Apron Cast actually do? just for floating point and rounding? *)
-                | false
-                | exception Cilfacade.TypeOfError _ (* typeOf inner e, not outer exp *)
-                | exception Invalid_argument _ -> (* get_ikind in is_cast_injective *)
-                  raise (Unsupported_CilExp (Cast_not_injective t))
-              end
-            | _ ->
-              raise (Unsupported_CilExp Exp_not_supported)
-          in
-          if not no_ov then (
-            let (type_min, type_max) = IntDomain.Size.range ik in
-            let texpr1 = Texpr1.of_expr env expr in
-            match Bounds.bound_texpr d texpr1 with
-            | Some min, Some max when Z.compare type_min min <= 0 && Z.compare max type_max <= 0 -> ()
-            | min_opt, max_opt ->
-              if M.tracing then M.trace "apron" "may overflow: %a (%a, %a)\n" CilType.Exp.pretty exp (Pretty.docOpt (IntOps.BigIntOps.pretty ())) min_opt (Pretty.docOpt (IntOps.BigIntOps.pretty ())) max_opt;
-              raise (Unsupported_CilExp Overflow)
-          );
-          expr
-        | exception (Cilfacade.TypeOfError _ as e)
-        | exception (Invalid_argument _ as e) ->
-          raise (Unsupported_CilExp (Exp_typeOf e))
+            failwith "texpr1_expr_of_cil_exp: globals must be replaced with temporary locals"
+        | Const (CInt (i, _, _)) ->
+          Cst (Coeff.s_of_mpqf (Mpqf.of_mpz (Z_mlgmpidl.mpz_of_z i)))
+        | exp ->
+          match Cilfacade.get_ikind_exp exp with
+          | ik ->
+            let expr =
+              let simplify e =
+                let ikind = try (Cilfacade.get_ikind_exp e) with Invalid_argument _ -> raise (Unsupported_CilExp Exp_not_supported)   in
+                let simp = query e ikind in
+                let const = IntDomain.IntDomTuple.to_int @@ IntDomain.IntDomTuple.cast_to ikind simp in
+                if M.tracing then M.trace "relation" "texpr1_expr_of_cil_exp/simplify: %a -> %a\n" d_plainexp e IntDomain.IntDomTuple.pretty simp;
+                BatOption.map_default (fun c -> Const (CInt (c, ikind, None))) e const
```

Comment:
This is basically constant folding, in case some analysis (including or ourselves) has a constant evaluation for some subexpression.

Reply:
If the same analysis can constant fold a subexpression, it should just be able to do that on its own directly using all the pre-existing logic as well.
Inter-analysis subexpression constant folding is certainly more interesting, but do we have an example where it's actually necessary now?

Besides the precision aspect, there is a related performance gotcha with this sort of subexpression evaluation queries that appeared when we integrated Apron analysis into the modern framework.
Namely, in the common case where every subexpression doesn't fold to a constant, work is doubled at this branch and level in the evaluation tree. Because the `EvalInt` query made here is also answered by the analysis itself, so it processes the subexpression once and the result is thrown away here because it's not a constant. Then we recurse into the very same subexpressions as the query did, finally getting the non-constant result (the second time).
Now, this is repeated at all subexpressions, but also in all duplicated subevaluations in each one, which degrades from linear into super-exponential complexity, if I remember correctly.

I believe this is the reason why `ana.base.eval.deep-query` option exists, needed for actual benchmarks for the relational traces paper. So these things occur in real-world programs as well. Although we first observed it on some SV-COMP task with a particularly large expression I think.

Reply:
As there is no performance penalty observable for now, I'll mark this as resolved. We will see if it has a huge impact on the relational tests once I run the benchmarks for those for my thesis.